File: snippets/java/src/main/java/develop/TracingOtel.java
package develop;

// <start_here>
import dev.restate.sdk.HandlerRunner;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.interceptor.opentelemetry.OpenTelemetryInterceptorFactory;
import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.sdk.OpenTelemetrySdk;

class MyOtelApp {
  public static void main(String[] args) {
    OpenTelemetry openTelemetry = OpenTelemetrySdk.builder().build(); // your OpenTelemetry instance
    var otel = new OpenTelemetryInterceptorFactory(openTelemetry);

    var endpoint =
        Endpoint.bind(
                new MyService(),
                new HandlerRunner.Options()
                    .addHandlerInterceptorFactory(otel)
                    .addRunInterceptorFactory(otel))
            .build();
  }
}
// <end_here>

File: snippets/kotlin/src/main/kotlin/develop/TracingOtel.kt
package develop

// <start_here>
import dev.restate.sdk.interceptor.opentelemetry.kotlin.OpenTelemetryInterceptorFactory
import dev.restate.sdk.kotlin.HandlerRunner
import dev.restate.sdk.kotlin.endpoint.endpoint
import io.opentelemetry.api.OpenTelemetry
import io.opentelemetry.sdk.OpenTelemetrySdk

fun otelTracedEndpoint() {
  val openTelemetry: OpenTelemetry =
      OpenTelemetrySdk.builder().build() // your OpenTelemetry instance
  val otel = OpenTelemetryInterceptorFactory(openTelemetry)

  val endpoint = endpoint {
    bind(
        MyService(),
        HandlerRunner.Options(
            handlerInterceptorFactories = mutableListOf(otel),
            runInterceptorFactories = mutableListOf(otel),
        ),
    )
  }
}
// <end_here>
